### ISSUE TITLE
Network/data structure discussion

### ISSUE THREAD

**Issue by wdwatkins** [USGS-R/lake-temperature-neural-networks#5]

**Comment by wdwatkins** [comment]
> If each depth in the lake had its own output node, then all depths could talk to one another within the neural network, and maybe the network would figure out more about the acceptable depth structures even before we get to adding the physics-based loss function? If each depth in the lake had its own output node, then all depths could talk to one another within the neural network, and maybe the network would figure out more about the acceptable depth structures even before we get to adding the physics-based loss function?

I like the concept of this, although I think there would be complications with different number of observations in each profile.  I think we would have to standardize the number of observations, or can you have input nodes be 'dead' when there are fewer inputs?  For lakes with buoys or regular sampling programs, I suppose the data should be pretty regular.

If we can do either other of those though, it should be relatively simple to implement --- would just involve reshaping the data to a wider format.  

Generalizing the network between lakes could be tricky too — the network would probably need to get wider for deeper lakes, and vice versa.  This could be doable though.

**Comment by aappling-usgs** [comment]
All good points!

Regarding 'dead' observations (you do mean observations and not inputs, right?): My intuition says there should be a way to have some output nodes (depths) for which there are no observations sometimes, given that it's possible to do dropout regularization where you just knock out whole nodes willy-nilly on each fitting iteration. But I don't know what the mechanics of it look like yet, just suspect it's possible.

I've also been wondering about whether this approach would cause problems for variable depths, and I don't think so. As long as we're training and predicting on just one lake at a time, yeah, we just adjust the structure of the model for that lake, and maybe anticipate different training times. And hopefully lakes don't change in depth over time enough to get in the way for this approach, or if they do, maybe we can again just treat the shortening/lengthening of the depth profile as "missing" observations / "dead" nodes?

**Comment by wdwatkins** [comment]
Right, I suppose there wouldn't be any reason to need to change what depths of GLM output are inputs.  The output nodes would just need to output the temps at the same depths, and some of those output nodes could be 'dead'.  We might have to limit the inputs to depths that are shallower than the minimum seasonal depth, so that there aren't NAs being fed in.  

I suppose it wouldn't be too difficult to automate deciding which depths have enough observations to be used as inputs from GLM.  For Mendota at least, the vast majority of observations are at whole or half meter intervals, so that wouldn't be an issue.  Other lakes might be trickier that have more of a grab bag of observations.
